diff --git a/spacy/cli/_util.py b/spacy/cli/_util.py
index fa41e6a08e0..1b42b5254bf 100644
--- a/spacy/cli/_util.py
+++ b/spacy/cli/_util.py
@@ -20,7 +20,7 @@
 import srsly
 import typer
 from click import NoSuchOption
-from click.parser import split_arg_string
+from click.shell_completion import split_arg_string
 from thinc.api import Config, ConfigValidationError, require_gpu
 from thinc.util import gpu_is_available
 from typer.main import get_command
