diff --git a/pydantic/mypy.py b/pydantic/mypy.py
index 475aa3e550..1bacb37d8d 100644
--- a/pydantic/mypy.py
+++ b/pydantic/mypy.py
@@ -7,6 +7,8 @@
 from configparser import ConfigParser
 from typing import Any
 
+# To avoid import error (https://github.com/python/mypy/issues/17726):
+import mypy.types  # noqa: F401
 from mypy.errorcodes import ErrorCode
 from mypy.expandtype import expand_type, expand_type_by_instance
 from mypy.nodes import (
