diff --git a/httpx/_config.py b/httpx/_config.py
index 25656b81ff..9318de3c13 100644
--- a/httpx/_config.py
+++ b/httpx/_config.py
@@ -1,12 +1,14 @@
 from __future__ import annotations
 
-import ssl
 import typing
 
 from ._models import Headers
 from ._types import HeaderTypes, TimeoutTypes
 from ._urls import URL
 
+if typing.TYPE_CHECKING:
+    import ssl  # pragma: no cover
+
 __all__ = ["Limits", "Proxy", "Timeout", "create_ssl_context"]
 
 
