diff --git a/doc/source/whatsnew/v3.1.0.rst b/doc/source/whatsnew/v3.1.0.rst
index 90adf51c0cae4..8a1d429205eff 100644
--- a/doc/source/whatsnew/v3.1.0.rst
+++ b/doc/source/whatsnew/v3.1.0.rst
@@ -129,7 +129,7 @@ Deprecations
   :meth:`DataFrame.prod` (and their :class:`Series` equivalents); this will raise in
   a future version of pandas (:issue:`53098`)
 - Deprecated the inference of ``datetime64`` dtype from data containing ``datetime.date`` objects when used in comparisons or :meth:`Index.equals` with :class:`DatetimeIndex`. Use :func:`pandas.to_datetime` to explicitly convert to ``datetime64`` instead (:issue:`65056`)
--
+- Deprecated ``inplace`` keyword :meth:`DataFrame.rename`. This keyword will be removed in a future version (:issue:`63207`); see also `PDEP-8 <https://pandas.pydata.org/pdeps/0008-inplace-methods-in-pandas.html>`_
 
 .. ---------------------------------------------------------------------------
 .. _whatsnew_310.performance:
diff --git a/pandas/core/frame.py b/pandas/core/frame.py
index d121a372d2fe4..42bad658b02f0 100644
--- a/pandas/core/frame.py
+++ b/pandas/core/frame.py
@@ -6242,7 +6242,7 @@ def rename(
         columns: Renamer | None = ...,
         axis: Axis | None = ...,
         copy: bool | lib.NoDefault = lib.no_default,
-        inplace: bool = ...,
+        inplace: bool | lib.NoDefault = lib.no_default,
         level: Level = ...,
         errors: IgnoreRaise = ...,
     ) -> DataFrame | None: ...
@@ -6255,7 +6255,7 @@ def rename(
         columns: Renamer | None = None,
         axis: Axis | None = None,
         copy: bool | lib.NoDefault = lib.no_default,
-        inplace: bool = False,
+        inplace: bool | lib.NoDefault = lib.no_default,
         level: Level | None = None,
         errors: IgnoreRaise = "ignore",
     ) -> DataFrame | None:
@@ -6300,6 +6300,14 @@ def rename(
         inplace : bool, default False
             Whether to modify the DataFrame rather than creating a new one.
             If True then value of copy is ignored.
+
+            .. deprecated:: 3.1.0
+
+                This keyword is deprecated and will be removed in pandas 4.0.
+                See `PDEP-8 In-place methods in pandas
+                <https://pandas.pydata.org/pdeps/0008-inplace-methods-in-pandas.html>`__
+                for more details.
+
         level : int or level name, default None
             In case of a MultiIndex, only rename labels in the specified
             level.
@@ -6377,7 +6385,22 @@ def rename(
         2  2  5
         4  3  6
         """
+
+        if inplace is not lib.no_default:
+            # GH#63207
+            warnings.warn(
+                "The inplace keyword in DataFrame.rename is "
+                "deprecated and will be removed in a future version. "
+                "See `PDEP-8 for more details:"
+                "https://pandas.pydata.org/pdeps/0008-inplace-methods-in-pandas.html",
+                Pandas4Warning,
+                stacklevel=find_stack_level(),
+            )
+        else:
+            inplace = False
+
         self._check_copy_deprecation(copy)
+
         return super()._rename(
             mapper=mapper,
             index=index,
diff --git a/pandas/tests/frame/methods/test_rename.py b/pandas/tests/frame/methods/test_rename.py
index 5271f304e9763..ecb026ee1fede 100644
--- a/pandas/tests/frame/methods/test_rename.py
+++ b/pandas/tests/frame/methods/test_rename.py
@@ -4,6 +4,8 @@
 import numpy as np
 import pytest
 
+from pandas.errors import Pandas4Warning
+
 from pandas import (
     NA,
     DataFrame,
@@ -175,6 +177,7 @@ def test_rename_nocopy(self, float_frame):
         renamed.loc[:, "foo"] = 1.0
         assert not (float_frame["C"] == 1.0).all()
 
+    @pytest.mark.filterwarnings("ignore:The inplace keyword in DataFrame.rename is")
     def test_rename_inplace(self, float_frame):
         float_frame.rename(columns={"C": "foo"})
         assert "C" in float_frame
@@ -191,6 +194,19 @@ def test_rename_inplace(self, float_frame):
         # Used to be id(float_frame["foo"]) != c_id, but flaky in the CI
         assert float_frame["foo"] is not c_values
 
+    def test_rename_inplace_depr(self, float_frame):
+        msg = "The inplace keyword in DataFrame.rename is deprecated"
+
+        # uses keyword, sets to true, warning
+        with tm.assert_produces_warning(Pandas4Warning, match=msg):
+            float_frame.rename(columns={"C": "foo"}, inplace=True)
+        # uses keyword, sets to false, warning
+        with tm.assert_produces_warning(Pandas4Warning, match=msg):
+            float_frame.rename(columns={"C": "foo"}, inplace=False)
+        # does not use keyword, no warning
+        with tm.assert_produces_warning(False):
+            float_frame.rename(columns={"C": "foo"})
+
     def test_rename_bug(self):
         # GH 5344
         # rename set ref_locs, and set_index was not resetting
@@ -471,6 +487,7 @@ def test_rename_tuple_columns_not_multiindex(self):
             result.columns, Index([(1, 1), (2, 2)], tupleize_cols=False)
         )
 
+    @pytest.mark.filterwarnings("ignore:The inplace keyword in DataFrame.rename is")
     def test_rename_non_unique_index_series(self):
         # GH#58621
         df = DataFrame({"A": [1, 2, 3], "B": [4, 5, 6], "C": [7, 8, 9]})
diff --git a/pandas/tests/frame/test_api.py b/pandas/tests/frame/test_api.py
index 4eaf112c9ffa9..e1a5fa98c942e 100644
--- a/pandas/tests/frame/test_api.py
+++ b/pandas/tests/frame/test_api.py
@@ -226,6 +226,7 @@ def test_deepcopy(self, float_frame):
         cp.loc[0, "A"] = 10
         assert not float_frame.equals(cp)
 
+    @pytest.mark.filterwarnings("ignore:The inplace keyword in DataFrame.rename is")
     def test_inplace_return_self(self):
         # GH 1893
 
diff --git a/pandas/tests/generic/test_duplicate_labels.py b/pandas/tests/generic/test_duplicate_labels.py
index 2e629bef109eb..ad71788628efa 100644
--- a/pandas/tests/generic/test_duplicate_labels.py
+++ b/pandas/tests/generic/test_duplicate_labels.py
@@ -375,6 +375,7 @@ def test_dataframe_insert_raises():
         (operator.methodcaller("rename", lambda x: x, inplace=True), False),
     ],
 )
+@pytest.mark.filterwarnings("ignore:The inplace keyword in DataFrame.rename is")
 def test_inplace_raises(method, frame_only):
     df = pd.DataFrame({"A": [0, 0], "B": [1, 2]}).set_flags(
         allows_duplicate_labels=False
