diff --git a/doc/source/whatsnew/v3.1.0.rst b/doc/source/whatsnew/v3.1.0.rst
index 174db226ec648..19ef0c4a4a58d 100644
--- a/doc/source/whatsnew/v3.1.0.rst
+++ b/doc/source/whatsnew/v3.1.0.rst
@@ -135,7 +135,7 @@ Deprecations
   :meth:`DataFrame.prod` (and their :class:`Series` equivalents); this will raise in
   a future version of pandas (:issue:`53098`)
 - Deprecated the inference of ``datetime64`` dtype from data containing ``datetime.date`` objects when used in comparisons or :meth:`Index.equals` with :class:`DatetimeIndex`. Use :func:`pandas.to_datetime` to explicitly convert to ``datetime64`` instead (:issue:`65056`)
-- Deprecated ``inplace`` keyword :meth:`DataFrame.rename`. This keyword will be removed in a future version (:issue:`63207`); see also `PDEP-8 <https://pandas.pydata.org/pdeps/0008-inplace-methods-in-pandas.html>`_
+- Deprecated ``inplace`` keyword for :meth:`DataFrame.rename` and :meth:`DataFrame.drop`. This keyword will be removed in a future version (:issue:`63207`); see also `PDEP-8 <https://pandas.pydata.org/pdeps/0008-inplace-methods-in-pandas.html>`_
 
 .. ---------------------------------------------------------------------------
 .. _whatsnew_310.performance:
diff --git a/pandas/core/frame.py b/pandas/core/frame.py
index 684f19fa3c89c..4c1a6762f537c 100644
--- a/pandas/core/frame.py
+++ b/pandas/core/frame.py
@@ -6060,7 +6060,7 @@ def drop(
         index: IndexLabel | ListLike = ...,
         columns: IndexLabel | ListLike = ...,
         level: Level = ...,
-        inplace: bool = ...,
+        inplace: bool | lib.NoDefault = lib.no_default,
         errors: IgnoreRaise = ...,
     ) -> DataFrame | None: ...
 
@@ -6072,7 +6072,7 @@ def drop(
         index: IndexLabel | ListLike = None,
         columns: IndexLabel | ListLike = None,
         level: Level | None = None,
-        inplace: bool = False,
+        inplace: bool | lib.NoDefault = lib.no_default,
         errors: IgnoreRaise = "raise",
     ) -> DataFrame | None:
         """
@@ -6103,6 +6103,14 @@ def drop(
         inplace : bool, default False
             If False, return a copy. Otherwise, do operation
             in place and return None.
+
+            .. deprecated:: 3.1.0
+
+                This keyword is deprecated and will be removed in pandas 4.0.
+                See `PDEP-8 In-place methods in pandas
+                <https://pandas.pydata.org/pdeps/0008-inplace-methods-in-pandas.html>`__
+                for more details.
+
         errors : {'ignore', 'raise'}, default 'raise'
             If 'ignore', suppress error and only existing labels are
             dropped.
@@ -6222,6 +6230,19 @@ def drop(
         falcon  speed   320.0   250.0
                 weight  1.0     0.8
         """
+        if inplace is not lib.no_default:
+            # GH#63207
+            warnings.warn(
+                "The inplace keyword in DataFrame.drop is deprecated "
+                "and will be removed in a future version. "
+                "See PDEP-8 for more details:"
+                "https://pandas.pydata.org/pdeps/0008-inplace-methods-in-pandas.html",
+                Pandas4Warning,
+                stacklevel=find_stack_level(),
+            )
+        else:
+            inplace = False
+
         return super().drop(
             labels=labels,
             axis=axis,
@@ -6418,7 +6439,7 @@ def rename(
             warnings.warn(
                 "The inplace keyword in DataFrame.rename is "
                 "deprecated and will be removed in a future version. "
-                "See `PDEP-8 for more details:"
+                "See PDEP-8 for more details:"
                 "https://pandas.pydata.org/pdeps/0008-inplace-methods-in-pandas.html",
                 Pandas4Warning,
                 stacklevel=find_stack_level(),
diff --git a/pandas/core/generic.py b/pandas/core/generic.py
index 5a2c88f5aac2c..f55a6d366e2b5 100644
--- a/pandas/core/generic.py
+++ b/pandas/core/generic.py
@@ -1855,7 +1855,7 @@ def _drop_labels_or_levels(self, keys, axis: AxisInt = 0):
 
             # Handle dropping columns labels
             if labels_to_drop:
-                dropped.drop(labels_to_drop, axis=1, inplace=True)
+                dropped = dropped.drop(labels_to_drop, axis=1)
         else:
             # Handle dropping column levels
             if levels_to_drop:
@@ -1869,7 +1869,7 @@ def _drop_labels_or_levels(self, keys, axis: AxisInt = 0):
 
             # Handle dropping index labels
             if labels_to_drop:
-                dropped.drop(labels_to_drop, axis=0, inplace=True)
+                dropped = dropped.drop(labels_to_drop, axis=0)
 
         return dropped
 
diff --git a/pandas/tests/frame/methods/test_drop.py b/pandas/tests/frame/methods/test_drop.py
index 70211af31b564..3630a6905e96a 100644
--- a/pandas/tests/frame/methods/test_drop.py
+++ b/pandas/tests/frame/methods/test_drop.py
@@ -3,6 +3,8 @@
 import numpy as np
 import pytest
 
+from pandas.errors import Pandas4Warning
+
 import pandas as pd
 from pandas import (
     DataFrame,
@@ -65,6 +67,7 @@ def test_drop_with_non_unique_datetime_index_and_invalid_keys():
 
 
 class TestDataFrameDrop:
+    @pytest.mark.filterwarnings("ignore:The inplace keyword in DataFrame.drop is")
     def test_drop_names(self):
         df = DataFrame(
             [[1, 2, 3], [3, 4, 5], [5, 6, 7]],
@@ -112,6 +115,7 @@ def test_drop_names(self):
         expected = Index(["a", "b", "c"], name="first")
         tm.assert_index_equal(dropped.index, expected)
 
+    @pytest.mark.filterwarnings("ignore:The inplace keyword in DataFrame.drop is")
     def test_drop(self):
         simple = DataFrame({"A": [1, 2, 3, 4], "B": [0, 1, 2, 3]})
         tm.assert_frame_equal(simple.drop("A", axis=1), simple[["B"]])
@@ -462,7 +466,9 @@ def test_inplace_drop_and_operation(self, operation, inplace):
         df["y"] = range(5)
         y = df["y"]
 
-        with tm.assert_produces_warning(None):
+        # For GH 63207 since pytest.mark.filterwarning did not filter warning here
+        msg = "The inplace keyword in DataFrame.drop is deprecated"
+        with tm.assert_produces_warning(Pandas4Warning, match=msg):
             if inplace:
                 df.drop("y", axis=1, inplace=inplace)
             else:
@@ -517,6 +523,7 @@ def test_drop_with_duplicate_columns2(self):
         result = df2.drop("C", axis=1)
         tm.assert_frame_equal(result, expected)
 
+    @pytest.mark.filterwarnings("ignore:The inplace keyword in DataFrame.drop is")
     def test_drop_inplace_no_leftover_column_reference(self):
         # GH 13934
         df = DataFrame({"a": [1, 2, 3]}, columns=Index(["a"], dtype="object"))
@@ -553,3 +560,21 @@ def test_drop_parse_strings_datetime_index(self):
         result = df.drop("2000-01-03", axis=0)
         expected = DataFrame({"a": [2], "b": [2]}, index=[Timestamp("2000-01-04")])
         tm.assert_frame_equal(result, expected)
+
+    @pytest.mark.filterwarnings("ignore:The inplace keyword in DataFrame.drop is")
+    def test_drop_inplace_depr(self):
+        # GH 63207
+        df = DataFrame(
+            {"a": [1, 2, 3], "b": [4, 5, 6], "c": [7, 8, 9], "d": [10, 11, 12]}
+        )
+        msg = "The inplace keyword in DataFrame.drop is deprecated"
+
+        # uses keyword, sets to true, warning
+        with tm.assert_produces_warning(Pandas4Warning, match=msg):
+            df.drop(columns=["a"], inplace=True)
+        # uses keyword, sets to false, warning
+        with tm.assert_produces_warning(Pandas4Warning, match=msg):
+            df.drop(columns=["b"], inplace=False)
+        # does not use keyword, no warning
+        with tm.assert_produces_warning(False):
+            df.drop(columns=["c"])
diff --git a/pandas/tests/frame/methods/test_dropna.py b/pandas/tests/frame/methods/test_dropna.py
index 11893d7fac1a4..e8d1ce5f9307a 100644
--- a/pandas/tests/frame/methods/test_dropna.py
+++ b/pandas/tests/frame/methods/test_dropna.py
@@ -129,6 +129,7 @@ def test_dropna(self):
         with pytest.raises(ValueError, match=msg):
             df.dropna(axis=3)
 
+    @pytest.mark.filterwarnings("ignore:The inplace keyword in DataFrame.drop is")
     def test_drop_and_dropna_caching(self):
         # tst that cacher updates
         original = Series([1, 2, np.nan], name="A")
diff --git a/pandas/tests/generic/test_frame.py b/pandas/tests/generic/test_frame.py
index c2d24cceeab0c..f37284eb7f63a 100644
--- a/pandas/tests/generic/test_frame.py
+++ b/pandas/tests/generic/test_frame.py
@@ -153,6 +153,7 @@ def test_deepcopy_empty(self):
 
 # formerly in Generic but only test DataFrame
 class TestDataFrame2:
+    @pytest.mark.filterwarnings("ignore:The inplace keyword in DataFrame.drop is")
     @pytest.mark.parametrize("value", [1, "True", [1, 2, 3], 5.0])
     def test_validate_bool_args(self, value):
         df = DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})
diff --git a/pandas/tests/indexing/multiindex/test_indexing_slow.py b/pandas/tests/indexing/multiindex/test_indexing_slow.py
index 3a3e23e713f07..2aaae00fa92a7 100644
--- a/pandas/tests/indexing/multiindex/test_indexing_slow.py
+++ b/pandas/tests/indexing/multiindex/test_indexing_slow.py
@@ -101,8 +101,7 @@ def test_multiindex_get_loc(request, lexsort_depth, keys, frame_fixture, cols):
             right = df[mask].copy(deep=False)
 
             if i + 1 != len(key):  # partial key
-                return_value = right.drop(cols[: i + 1], axis=1, inplace=True)
-                assert return_value is None
+                right = right.drop(cols[: i + 1], axis=1)
                 return_value = right.set_index(cols[i + 1 : -1], inplace=True)
                 assert return_value is None
                 tm.assert_frame_equal(mi.loc[key[: i + 1]], right)
diff --git a/pandas/tests/plotting/test_series.py b/pandas/tests/plotting/test_series.py
index 6b572ae65c264..542688c2f1228 100644
--- a/pandas/tests/plotting/test_series.py
+++ b/pandas/tests/plotting/test_series.py
@@ -844,6 +844,7 @@ def test_custom_business_day_freq(self):
 
         _check_plot_works(s.plot)
 
+    @pytest.mark.filterwarnings("ignore:The inplace keyword in DataFrame.drop is")
     @pytest.mark.xfail(
         reason="GH#24426, see also "
         "github.com/pandas-dev/pandas/commit/"
diff --git a/pandas/tests/series/accessors/test_str_accessor.py b/pandas/tests/series/accessors/test_str_accessor.py
index ff530459b78fb..9970cbc8c3693 100644
--- a/pandas/tests/series/accessors/test_str_accessor.py
+++ b/pandas/tests/series/accessors/test_str_accessor.py
@@ -19,6 +19,7 @@ def test_str_attribute(self):
         with pytest.raises(AttributeError, match=msg):
             ser.str.repeat(2)
 
+    @pytest.mark.filterwarnings("ignore:The inplace keyword in DataFrame.drop is")
     def test_str_accessor_updates_on_inplace(self):
         ser = Series(list("abc"))
         return_value = ser.drop([0], inplace=True)
