diff --git a/doc/modules/tree.rst b/doc/modules/tree.rst
index d0335cd2f6b89..e1140188ca7e1 100644
--- a/doc/modules/tree.rst
+++ b/doc/modules/tree.rst
@@ -639,16 +639,12 @@ Note that it is 3–6× slower to fit than the MSE criterion as of version 1.8.
 Missing Values Support
 ======================
 
-:class:`DecisionTreeClassifier`, :class:`DecisionTreeRegressor`
-have built-in support for missing values using `splitter='best'`, where
-the splits are determined in a greedy fashion.
+:class:`DecisionTreeClassifier`, :class:`DecisionTreeRegressor` and
 :class:`ExtraTreeClassifier`, and :class:`ExtraTreeRegressor` have built-in
-support for missing values for `splitter='random'`, where the splits
-are determined randomly. For more details on how the splitter differs on
-non-missing values, see the :ref:`Forest section <forest>`.
+support for missing values except for sparse inputs.
 
 First we will describe how :class:`DecisionTreeClassifier`, :class:`DecisionTreeRegressor`
-handle missing-values in the data.
+(``splitter='best'``) handle missing-values in the data.
 
 For each potential threshold on the non-missing data, the splitter will evaluate
 the split with all the missing values going to the left node or the right node.
@@ -700,12 +696,11 @@ Decisions are made as follows:
     >>> tree.predict(X_test)
     array([1])
 
-:class:`ExtraTreeClassifier`, and :class:`ExtraTreeRegressor` handle missing values
-in a slightly different way. When splitting a node, a random threshold will be chosen
-to split the non-missing values on. Then the non-missing values will be sent to the
-left and right child based on the randomly selected threshold, while the missing
-values will also be randomly sent to the left or right child. This is repeated for
-every feature considered at each split. The best split among these is chosen.
+:class:`ExtraTreeClassifier`, and :class:`ExtraTreeRegressor` (``splitter='random'``)
+handle missing values in a slightly different way. When splitting a node, a random
+threshold will be chosen to split the non-missing values on, while the missing values
+are assigned together to one randomly chosen side. This is repeated for every
+feature considered at each split. The best split among these is chosen.
 
 During prediction, the treatment of missing-values is the same as that of the
 decision tree:
diff --git a/sklearn/tree/_classes.py b/sklearn/tree/_classes.py
index 313072da4a9cc..5945fce431b22 100644
--- a/sklearn/tree/_classes.py
+++ b/sklearn/tree/_classes.py
@@ -25,9 +25,8 @@
     clone,
     is_classifier,
 )
-from sklearn.tree import _criterion, _splitter
+from sklearn.tree import _criterion, _splitter  # type: ignore[attr-defined]
 from sklearn.tree._criterion import Criterion
-from sklearn.tree._splitter import Splitter
 from sklearn.tree._tree import (
     BestFirstTreeBuilder,
     DepthFirstTreeBuilder,
@@ -381,9 +380,6 @@ def _fit(
             # might be shared and modified concurrently during parallel fitting
             criterion = copy.deepcopy(criterion)
 
-        SPLITTERS = SPARSE_SPLITTERS if issparse(X) else DENSE_SPLITTERS
-
-        splitter = self.splitter
         if self.monotonic_cst is None:
             monotonic_cst = None
         else:
@@ -423,15 +419,15 @@ def _fit(
                 # *positive class*, all signs must be flipped.
                 monotonic_cst *= -1
 
-        if not isinstance(self.splitter, Splitter):
-            splitter = SPLITTERS[self.splitter](
-                criterion,
-                self.max_features_,
-                min_samples_leaf,
-                min_weight_leaf,
-                random_state,
-                monotonic_cst,
-            )
+        SPLITTERS = SPARSE_SPLITTERS if issparse(X) else DENSE_SPLITTERS
+        splitter = SPLITTERS[self.splitter](
+            criterion,
+            self.max_features_,
+            min_samples_leaf,
+            min_weight_leaf,
+            random_state,
+            monotonic_cst,
+        )
 
         if is_classifier(self):
             self.tree_ = Tree(self.n_features_in_, self.n_classes_, self.n_outputs_)
@@ -1718,15 +1714,6 @@ def __init__(
             monotonic_cst=monotonic_cst,
         )
 
-    def __sklearn_tags__(self):
-        tags = super().__sklearn_tags__()
-        # XXX: nan values are only accepted in dense arrays, but we set this for
-        # common test to pass, specifically: check_estimators_nan_inf
-        allow_nan = self.splitter == "random"
-        tags.classifier_tags.multi_label = True
-        tags.input_tags.allow_nan = allow_nan
-        return tags
-
 
 class ExtraTreeRegressor(DecisionTreeRegressor):
     """An extremely randomized tree regressor.
@@ -1972,11 +1959,3 @@ def __init__(
             ccp_alpha=ccp_alpha,
             monotonic_cst=monotonic_cst,
         )
-
-    def __sklearn_tags__(self):
-        tags = super().__sklearn_tags__()
-        # XXX: nan values are only accepted in dense arrays, but we set this for
-        # common test to pass, specifically: check_estimators_nan_inf
-        allow_nan = self.splitter == "random"
-        tags.input_tags.allow_nan = allow_nan
-        return tags
