diff --git a/examples/svm/plot_iris_svc.py b/examples/svm/plot_iris_svc.py
index 77259f9d1ea2c..cb16b69b85f91 100644
--- a/examples/svm/plot_iris_svc.py
+++ b/examples/svm/plot_iris_svc.py
@@ -9,8 +9,8 @@
 - Sepal length
 - Sepal width
 
-This example shows how to plot the decision surface for four SVM classifiers
-with different kernels.
+This example shows how to plot the decision surface and the support vectors for
+four SVM classifiers with different kernels.
 
 The linear models ``LinearSVC()`` and ``SVC(kernel='linear')`` yield slightly
 different decision boundaries. This can be a consequence of the following
@@ -27,7 +27,7 @@
 flexible non-linear decision boundaries with shapes that depend on the kind of
 kernel and its parameters.
 
-.. NOTE:: while plotting the decision function of classifiers for toy 2D
+.. NOTE:: While plotting the decision function of classifiers for toy 2D
    datasets can help get an intuitive understanding of their respective
    expressive power, be aware that those intuitions don't always generalize to
    more realistic high-dimensional problems.
@@ -38,18 +38,19 @@
 # SPDX-License-Identifier: BSD-3-Clause
 
 import matplotlib.pyplot as plt
+import numpy as np
 
 from sklearn import datasets, svm
 from sklearn.inspection import DecisionBoundaryDisplay
 
-# import some data to play with
+# Import some data to play with.
 iris = datasets.load_iris()
-# Take the first two features. We could avoid this by using a two-dim dataset
+# Take the first two features. We could avoid this by using a two-dim dataset.
 X = iris.data[:, :2]
 y = iris.target
 
-# we create an instance of SVM and fit out data. We do not scale our
-# data since we want to plot the support vectors
+# We create an instance of SVM and fit out data. We do not scale our
+# data since we want to plot the support vectors.
 C = 1.0  # SVM regularization parameter
 models = (
     svm.SVC(kernel="linear", C=C),
@@ -59,7 +60,7 @@
 )
 models = (clf.fit(X, y) for clf in models)
 
-# title for the plots
+# Title for the plots
 titles = (
     "SVC with linear kernel",
     "LinearSVC (linear kernel)",
@@ -71,20 +72,34 @@
 fig, sub = plt.subplots(2, 2)
 plt.subplots_adjust(wspace=0.4, hspace=0.4)
 
-X0, X1 = X[:, 0], X[:, 1]
-
 for clf, title, ax in zip(models, titles, sub.flatten()):
     disp = DecisionBoundaryDisplay.from_estimator(
         clf,
         X,
         response_method="predict",
-        cmap=plt.cm.coolwarm,
+        multiclass_colors="coolwarm",
         alpha=0.8,
         ax=ax,
         xlabel=iris.feature_names[0],
         ylabel=iris.feature_names[1],
     )
-    ax.scatter(X0, X1, c=y, cmap=plt.cm.coolwarm, s=20, edgecolors="k")
+
+    # Plot the support vectors.
+    # For LinearSVC we compute the support vectors from the decision function, see
+    # https://scikit-learn.org/dev/auto_examples/svm/plot_linearsvc_support_vectors.html
+    if hasattr(clf, "support_"):
+        support_vector_indices = clf.support_
+    else:
+        decision_function = clf.decision_function(X)
+        support_vector_indices = (np.abs(decision_function) <= 1 + 1e-15).nonzero()[0]
+    ax.scatter(
+        X[support_vector_indices, 0],
+        X[support_vector_indices, 1],
+        c=y[support_vector_indices],
+        cmap=plt.cm.coolwarm,
+        edgecolors="k",
+    )
+
     ax.set_xticks(())
     ax.set_yticks(())
     ax.set_title(title)
