diff --git a/changelogs/fragments/windows-compression.yml b/changelogs/fragments/windows-compression.yml
new file mode 100644
index 00000000000000..d6691a0adc81fb
--- /dev/null
+++ b/changelogs/fragments/windows-compression.yml
@@ -0,0 +1,5 @@
+minor_changes:
+  - >-
+    windows - Compress input data sent over for module execution to reduce the amount of data transferred per module
+    execution.
+    
diff --git a/lib/ansible/config/base.yml b/lib/ansible/config/base.yml
index d034f8c8120037..f16a071e185ac4 100644
--- a/lib/ansible/config/base.yml
+++ b/lib/ansible/config/base.yml
@@ -31,6 +31,16 @@ _ANSIBALLZ_PYDEVD_CONFIG:
   vars:
   - {name: _ansible_ansiballz_pydevd_config}
   version_added: '2.20'
+_ANSIBALLZ_PWSH_COMPRESSION:
+  name: Enable or disable compression used for PowerShell AnsiballZ data
+  description:
+    - Enables or disables compression used for PowerShell AnsiballZ data.
+    - This is for internal use only and may be removed without warning.
+  env:
+  - name: _ANSIBLE_ANSIBALLZ_PWSH_COMPRESSION
+  type: boolean
+  default: true
+  version_added: '2.22'
 _ANSIBLE_CONNECTION_PATH:
   env:
   - name: _ANSIBLE_CONNECTION_PATH
@@ -1663,7 +1673,7 @@ INTERPRETER_PYTHON:
   description:
   - Path to the Python interpreter to be used for module execution on remote targets, or an automatic discovery mode.
     Supported discovery modes are ``auto`` (the default), and ``auto_silent``.
-    All discovery modes match against an ordered list of well-known Python interpreter locations. 
+    All discovery modes match against an ordered list of well-known Python interpreter locations.
     The fallback behavior will issue a warning that the interpreter should be set explicitly (since interpreters
     installed later may change which one is used). This warning behavior can be disabled by setting ``auto_silent``.
     The ``auto_legacy*`` modes are removed.
diff --git a/lib/ansible/executor/powershell/exec_wrapper.ps1 b/lib/ansible/executor/powershell/exec_wrapper.ps1
index 4c2b27e60e0b54..5d40c6b5509dfc 100644
--- a/lib/ansible/executor/powershell/exec_wrapper.ps1
+++ b/lib/ansible/executor/powershell/exec_wrapper.ps1
@@ -5,6 +5,7 @@ using namespace System.Collections
 using namespace System.Collections.Generic
 using namespace System.Diagnostics.CodeAnalysis
 using namespace System.IO
+using namespace System.IO.Compression
 using namespace System.Linq
 using namespace System.Management.Automation
 using namespace System.Management.Automation.Language
@@ -48,6 +49,10 @@ param (
     [string]
     $PwshPath,
 
+    [Parameter()]
+    [switch]
+    $DecompressInput,
+
     [Parameter()]
     [PSObject]
     $ActionParameters
@@ -139,6 +144,10 @@ begin {
                 $encCommand
             )
 
+            # Switch parameters need to be converted to a boolean so they
+            # serialized properly in JSON
+            $PSBoundParameters['DecompressInput'] = $PSBoundParameters['DecompressInput'].IsPresent
+
             $execManifest = @{
                 name = 'exec_wrapper-respawn.ps1'
                 params = $PSBoundParameters
@@ -737,6 +746,7 @@ process {
         return
     }
 
+    $inputStream = $outputStream = $gzipStream = $null
     try {
         if ($actionPipeline) {
             # We received our manifest and started the action pipeline, redirect
@@ -753,6 +763,14 @@ process {
             if ($EncodeInputOutput) {
                 $jsonPipeline.Process([Encoding]::UTF8.GetString([Convert]::FromBase64String($InputObject)))
             }
+            elseif ($DecompressInput) {
+                $inputStream = [MemoryStream]::new([Convert]::FromBase64String($InputObject))
+                $gzipStream = [GZipStream]::new($inputStream, [CompressionMode]::Decompress)
+                $outputStream = [MemoryStream]::new()
+                $gzipStream.CopyTo($outputStream)
+
+                $jsonPipeline.Process([Encoding]::UTF8.GetString($outputStream.ToArray()))
+            }
             else {
                 $jsonPipeline.Process($InputObject)
             }
@@ -761,6 +779,11 @@ process {
     catch {
         Write-AnsibleErrorJson -ErrorRecord $_
     }
+    finally {
+        if ($inputStream) { $inputStream.Dispose() }
+        if ($outputStream) { $outputStream.Dispose() }
+        if ($gzipStream) { $gzipStream.Dispose() }
+    }
 }
 
 end {
diff --git a/lib/ansible/executor/powershell/module_manifest.py b/lib/ansible/executor/powershell/module_manifest.py
index 478d2574adf182..1d45a38e582076 100644
--- a/lib/ansible/executor/powershell/module_manifest.py
+++ b/lib/ansible/executor/powershell/module_manifest.py
@@ -26,6 +26,13 @@
 from ansible.plugins.loader import ps_module_utils_loader
 
 
+GZIP_AVAILABLE = True
+try:
+    import gzip
+except ImportError:
+    GZIP_AVAILABLE = False
+
+
 @dataclasses.dataclass(frozen=True)
 class _ExecManifest:
     scripts: dict[str, _ScriptInfo] = dataclasses.field(default_factory=dict)
@@ -565,6 +572,18 @@ def _get_bootstrap_input(
     :param temp_path: The temporary path to use for the scripts if needed.
     :return: The input for bootstrap_wrapper.ps1 as a byte string.
     """
+    decompress_input = False
+    exec_input = json.dumps(dataclasses.asdict(manifest)).encode()
+
+    # We provide an opt-out config option in case there are unforeseen issues
+    # with compression.
+    if GZIP_AVAILABLE and C.config.get_config_value("_ANSIBALLZ_PWSH_COMPRESSION", variables={}):
+        compressed_input = gzip.compress(exec_input)
+        compressed_b64 = base64.b64encode(compressed_input)
+        if len(compressed_b64) < len(exec_input):
+            exec_input = compressed_b64
+            decompress_input = True
+
     bootstrap_manifest = {
         'name': 'exec_wrapper',
         'script': _get_powershell_script("exec_wrapper.ps1").decode(),
@@ -573,12 +592,13 @@ def _get_bootstrap_input(
             'MinPSVersion': min_ps_version,
             'TempPath': temp_path,
             'PwshPath': pwsh_interpreter,
+            'DecompressInput': decompress_input,
         },
     }
 
-    bootstrap_input = json.dumps(bootstrap_manifest, ensure_ascii=True)
-    exec_input = json.dumps(dataclasses.asdict(manifest))
-    return f"{bootstrap_input}\n\0\0\0\0\n{exec_input}".encode()
+    bootstrap_input = json.dumps(bootstrap_manifest, ensure_ascii=True).encode()
+
+    return bootstrap_input + b"\n\0\0\0\0\n" + exec_input
 
 
 def _prepare_module_args(module_args: dict[str, t.Any], profile: str) -> dict[str, t.Any]:
