diff --git a/src/bokeh/application/handlers/code.py b/src/bokeh/application/handlers/code.py
index 1c6ae49486d..9cbaab2443d 100644
--- a/src/bokeh/application/handlers/code.py
+++ b/src/bokeh/application/handlers/code.py
@@ -44,6 +44,7 @@ def make_doc(doc: Document):
     Callable,
     ClassVar,
     Generator,
+    Sequence,
 )
 
 # Bokeh imports
@@ -88,7 +89,7 @@ class CodeHandler(Handler):
 
     _origin: ClassVar[str]
 
-    def __init__(self, *, source: str, filename: PathLike, argv: list[str] = [], package: ModuleType | None = None) -> None:
+    def __init__(self, *, source: str, filename: PathLike, argv: Sequence[str] = (), package: ModuleType | None = None) -> None:
         '''
 
         Args:
@@ -96,7 +97,7 @@ def __init__(self, *, source: str, filename: PathLike, argv: list[str] = [], pac
 
             filename (str) : a filename to use in any debugging or error output
 
-            argv (list[str], optional) : a list of string arguments to make
+            argv (Sequence[str], optional) : a sequence of string arguments to make
                 available as ``sys.argv`` when the code executes
 
         '''
diff --git a/src/bokeh/application/handlers/code_runner.py b/src/bokeh/application/handlers/code_runner.py
index 010a7cd4643..4c41e9f8ff6 100644
--- a/src/bokeh/application/handlers/code_runner.py
+++ b/src/bokeh/application/handlers/code_runner.py
@@ -28,7 +28,12 @@
 from contextlib import contextmanager
 from os.path import basename
 from threading import RLock
-from typing import TYPE_CHECKING, Callable, Generator
+from typing import (
+    TYPE_CHECKING,
+    Callable,
+    Generator,
+    Sequence,
+)
 
 # Bokeh imports
 from ...util.serialization import make_globally_unique_id
@@ -66,7 +71,7 @@ class CodeRunner:
     _permanent_error_detail: str | None
     _path: PathLike
     _source: str
-    _argv: list[str]
+    _argv: Sequence[str]
     _package: ModuleType | None
     ran: bool
 
@@ -74,7 +79,7 @@ class CodeRunner:
     _error: str | None
     _error_detail: str | None
 
-    def __init__(self, source: str, path: PathLike, argv: list[str], package: ModuleType | None = None) -> None:
+    def __init__(self, source: str, path: PathLike, argv: Sequence[str], package: ModuleType | None = None) -> None:
         '''
 
         Args:
@@ -84,8 +89,8 @@ def __init__(self, source: str, path: PathLike, argv: list[str], package: Module
             path (str) :
                 A filename to use in any debugging or error output
 
-            argv (list[str]) :
-                A list of string arguments to make available as ``sys.argv``
+            argv (Sequence[str]) :
+                A sequence of string arguments to make available as ``sys.argv``
                 when the code executes
 
             package (bool) :
@@ -246,7 +251,7 @@ def _hold_process_globals() -> Generator[None, None, None]:
         yield
 
 @contextmanager
-def _patch_process_state(path: PathLike, argv: list[str]) -> Generator[None, None, None]:
+def _patch_process_state(path: PathLike, argv: Sequence[str]) -> Generator[None, None, None]:
     # Simulate the sys.path behaviour described here:
     #
     # https://docs.python.org/2/library/sys.html#sys.path
diff --git a/src/bokeh/application/handlers/directory.py b/src/bokeh/application/handlers/directory.py
index 8aaed0b5824..e457e56096d 100644
--- a/src/bokeh/application/handlers/directory.py
+++ b/src/bokeh/application/handlers/directory.py
@@ -55,7 +55,7 @@
     exists,
     join,
 )
-from typing import TYPE_CHECKING, Any
+from typing import TYPE_CHECKING, Any, Sequence
 
 # External imports
 from jinja2 import Environment, FileSystemLoader, Template
@@ -107,12 +107,12 @@ class DirectoryHandler(Handler):
     _static: str | None
     _template: Template | None
 
-    def __init__(self, *, filename: PathLike, argv: list[str] = []) -> None:
+    def __init__(self, *, filename: PathLike, argv: Sequence[str] = ()) -> None:
         '''
         Keywords:
             filename (str) : a path to an application directory with either "main.py" or "main.ipynb"
 
-            argv (list[str], optional) : a list of string arguments to make available as sys.argv to main.py
+            argv (Sequence[str], optional) : a sequence of string arguments to make available as sys.argv to main.py
         '''
         super().__init__()
 
diff --git a/src/bokeh/application/handlers/notebook.py b/src/bokeh/application/handlers/notebook.py
index 9a83a77787f..dbd105b761f 100644
--- a/src/bokeh/application/handlers/notebook.py
+++ b/src/bokeh/application/handlers/notebook.py
@@ -30,7 +30,7 @@
 
 # Standard library imports
 import re
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Sequence
 
 # Bokeh imports
 from ...util.dependencies import import_required
@@ -67,7 +67,7 @@ class NotebookHandler(CodeHandler):
 
     _origin = "Notebook"
 
-    def __init__(self, *, filename: PathLike, argv: list[str] = [], package: ModuleType | None = None) -> None:
+    def __init__(self, *, filename: PathLike, argv: Sequence[str] = (), package: ModuleType | None = None) -> None:
         '''
 
         Keywords:
diff --git a/src/bokeh/application/handlers/script.py b/src/bokeh/application/handlers/script.py
index 045d96bdd19..c4c63b08f97 100644
--- a/src/bokeh/application/handlers/script.py
+++ b/src/bokeh/application/handlers/script.py
@@ -46,7 +46,7 @@
 #-----------------------------------------------------------------------------
 
 # Standard library imports
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Sequence
 
 # Bokeh imports
 from .code import CodeHandler
@@ -81,7 +81,7 @@ class ScriptHandler(CodeHandler):
 
     _origin = "Script"
 
-    def __init__(self, *, filename: PathLike, argv: list[str] = [], package: ModuleType | None = None) -> None:
+    def __init__(self, *, filename: PathLike, argv: Sequence[str] = (), package: ModuleType | None = None) -> None:
         '''
 
         Keywords:
diff --git a/src/bokeh/application/handlers/server_lifecycle.py b/src/bokeh/application/handlers/server_lifecycle.py
index 0efe9bdfe2f..a702d0924fa 100644
--- a/src/bokeh/application/handlers/server_lifecycle.py
+++ b/src/bokeh/application/handlers/server_lifecycle.py
@@ -23,7 +23,7 @@
 
 # Standard library imports
 import os
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Sequence
 
 # Bokeh imports
 from ...util.callback_manager import _check_callback
@@ -56,14 +56,14 @@ class ServerLifecycleHandler(LifecycleHandler):
 
     '''
 
-    def __init__(self, *, filename: PathLike, argv: list[str] = [], package: ModuleType | None = None) -> None:
+    def __init__(self, *, filename: PathLike, argv: Sequence[str] = (), package: ModuleType | None = None) -> None:
         '''
 
         Keyword Args:
             filename (str) : path to a module to load lifecycle callbacks from
 
-            argv (list[str], optional) : a list of string arguments to use as
-                ``sys.argv`` when the callback code is executed. (default: [])
+            argv (Sequence[str], optional) : a sequence of string arguments to use as
+                ``sys.argv`` when the callback code is executed. (default: ())
 
         '''
         super().__init__()
diff --git a/src/bokeh/application/handlers/server_request_handler.py b/src/bokeh/application/handlers/server_request_handler.py
index 99301d6e4bc..9ebca14ed59 100644
--- a/src/bokeh/application/handlers/server_request_handler.py
+++ b/src/bokeh/application/handlers/server_request_handler.py
@@ -23,7 +23,7 @@
 
 # Standard library imports
 import os
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Sequence
 
 # Bokeh imports
 from ...util.callback_manager import _check_callback
@@ -58,14 +58,14 @@ class ServerRequestHandler(RequestHandler):
 
     _module: ModuleType
 
-    def __init__(self, *, filename: PathLike, argv: list[str] = [], package: ModuleType | None = None) -> None:
+    def __init__(self, *, filename: PathLike, argv: Sequence[str] = (), package: ModuleType | None = None) -> None:
         '''
 
         Keyword Args:
             filename (str) : path to a module to load request handler callbacks from
 
-            argv (list[str], optional) : a list of string arguments to use as
-                ``sys.argv`` when the callback code is executed. (default: [])
+            argv (Sequence[str], optional) : a sequence of string arguments to use as
+                ``sys.argv`` when the callback code is executed. (default: ())
 
         '''
         super().__init__()
diff --git a/src/bokeh/core/property/wrappers.py b/src/bokeh/core/property/wrappers.py
index bc12b865c1a..1121c045ef6 100644
--- a/src/bokeh/core/property/wrappers.py
+++ b/src/bokeh/core/property/wrappers.py
@@ -411,8 +411,8 @@ def __setitem__(self, i: str, y: Sequence[Any]) -> None:
     def __copy__(self) -> PropertyValueColumnData:
         return PropertyValueColumnData(dict(self))
 
-    def __deepcopy__(self, memodict: dict[Any, Any] = {}) -> PropertyValueColumnData:
-        return PropertyValueColumnData(copy.deepcopy(dict(self), memodict))
+    def __deepcopy__(self, memo: dict[Any, Any]) -> PropertyValueColumnData:
+        return PropertyValueColumnData(copy.deepcopy(dict(self), memo))
 
     # don't wrap with notify_owner --- notifies owners explicitly
     def update(self, *args: Any, **kwargs: Sequence[Any]) -> None:
diff --git a/src/bokeh/layouts.py b/src/bokeh/layouts.py
index ce9224557ef..fbffcbb9b8b 100644
--- a/src/bokeh/layouts.py
+++ b/src/bokeh/layouts.py
@@ -398,7 +398,9 @@ def assert_unique[V](values: list[V], name: ToolbarOptions) -> V | UndefinedType
 
     return gp
 
-# XXX https://github.com/python/mypy/issues/731
+type GridChild = UIElement | None | Sequence[GridChild]
+type GridChildren = Sequence[GridChild]
+
 @overload
 def grid(children: list[UIElement | list[UIElement | list[Any]]], *, sizing_mode: SizingModeType | None = ...) -> GridBox: ...
 @overload
@@ -412,7 +414,12 @@ def grid(children: list[UIElement | None], *, sizing_mode: SizingModeType | None
 @overload
 def grid(children: str, *, sizing_mode: SizingModeType | None = ...) -> GridBox: ...
 
-def grid(children: Any = [], sizing_mode: SizingModeType | None = None, nrows: int | None = None, ncols: int | None = None) -> GridBox:
+def grid(
+    children: str | GridChildren | Row | Column = (),
+    sizing_mode: SizingModeType | None = None,
+    nrows: int | None = None,
+    ncols: int | None = None,
+) -> GridBox:
     """
     Conveniently create a grid of layoutable objects.
 
@@ -549,15 +556,18 @@ def _flatten(layout: row | col | LayoutDOM) -> Grid:
         return GridBox(children=children)
 
     layout: row | col
-    if isinstance(children, list):
+    if isinstance(children, str):
+        raise NotImplementedError
+    elif isinstance(children, Sequence):
+        children = list(children)
         if nrows is not None or ncols is not None:
             N = len(children)
             if ncols is None:
                 ncols = math.ceil(N/nrows)
             layout = col([ row(children[i:i+ncols]) for i in range(0, N, ncols) ])
         else:
-            def traverse_list(children: list[LayoutDOM], level: int = 0):
-                if isinstance(children, list):
+            def traverse_list(children: Sequence[GridChild], level: int = 0):
+                if isinstance(children, Sequence) and not isinstance(children, str):
                     container = col if level % 2 == 0 else row
                     return container([ traverse_list(child, level+1) for child in children ])
                 else:
@@ -576,10 +586,8 @@ def traverse_layout(item: LayoutDOM, top_level: bool = False):
                 return item
 
         layout = traverse_layout(children, top_level=True)
-    elif isinstance(children, str):
-        raise NotImplementedError
     else:
-        raise ValueError("expected a list, string or model")
+        raise ValueError("expected a sequence, string, Row or Column")
 
     grid = flatten(layout)
 
diff --git a/src/bokeh/models/widgets/tables.py b/src/bokeh/models/widgets/tables.py
index 9704dc953e6..e4fb1d4be21 100644
--- a/src/bokeh/models/widgets/tables.py
+++ b/src/bokeh/models/widgets/tables.py
@@ -925,6 +925,8 @@ def from_data(data, columns=None, formatters={}, **kwargs) -> DataTable:
 
         """
 
+
+        formatters = formatters.copy()
         if isinstance(data, ColumnDataSource):
             source = data.clone()
         else:
diff --git a/src/bokeh/plotting/_renderer.py b/src/bokeh/plotting/_renderer.py
index 210b263204d..4b84efbb1f3 100644
--- a/src/bokeh/plotting/_renderer.py
+++ b/src/bokeh/plotting/_renderer.py
@@ -192,6 +192,7 @@ def pop_visuals(glyphclass: type[Glyph], props: Attrs, *, prefix: str = "", defa
         ultimate defaults in case those can't be deduced.
     """
     defaults = defaults.copy()
+    override_defaults = override_defaults.copy()
     defaults.setdefault('text_color', 'black')
     defaults.setdefault('hatch_color', 'black')
 
diff --git a/src/bokeh/sphinxext/_internal/example_handler.py b/src/bokeh/sphinxext/_internal/example_handler.py
index e048edff139..71ee319a3e5 100644
--- a/src/bokeh/sphinxext/_internal/example_handler.py
+++ b/src/bokeh/sphinxext/_internal/example_handler.py
@@ -59,7 +59,7 @@ class ExampleHandler(Handler):
 
     def __init__(self, source: str, filename: PathLike) -> None:
         super().__init__()
-        self._runner = CodeRunner(source, filename, [])
+        self._runner = CodeRunner(source, filename, ())
 
     def modify_document(self, doc: Document) -> None:
         if self.failed:
diff --git a/tests/unit/bokeh/core/property/test_wrappers__property.py b/tests/unit/bokeh/core/property/test_wrappers__property.py
index 458071b71ea..c7c600f4ef5 100644
--- a/tests/unit/bokeh/core/property/test_wrappers__property.py
+++ b/tests/unit/bokeh/core/property/test_wrappers__property.py
@@ -17,6 +17,7 @@
 #-----------------------------------------------------------------------------
 
 # Standard library imports
+import copy
 from unittest.mock import MagicMock, patch
 
 # External imports
@@ -581,7 +582,7 @@ def test_PropertyValueColumnData___copy__() -> None:
 
 def test_PropertyValueColumnData___deepcopy__() -> None:
     source = ColumnDataSource(data=dict(foo=[10]))
-    pvcd = source.data.__deepcopy__()
+    pvcd = copy.deepcopy(source.data)
     assert source.data == pvcd
     assert id(source.data) != id(pvcd)
     pvcd['foo'][0] = 20
